r/src/array_from_vector.cpp

```diff
@@ -868,6 +868,17 @@ std::shared_ptr<arrow::DataType> InferType(SEXP x) {
                                                      InferType(VECTOR_ELT(x, i)));
         }
         return std::make_shared<StructType>(std::move(fields));
+      } else {
+        R_xlen_t n = XLENGTH(x);
+        if (n == 0)
+          Rcpp::stop("received length-0 list");
+        std::shared_ptr<arrow::Type> element_type = InferType(VECTOR_ELT(x, 0));
```

Comment:
Sounds like you weren't able to compile this locally, so see the results from CI: https://github.com/apache/arrow/pull/6275/checks?check_run_id=406491019#step:5:2868

There are a few things to fix here.